posts/inside-rust/2023-09-25-polonius-update.md

```diff
@@ -0,0 +1,152 @@
+
+The motivating example for Polonius is the so-called ["Problem Case #3: conditional control flow across functions"](https://github.com/rust-lang/rfcs/blob/master/text/2094-nll.md#problem-case-3-conditional-control-flow-across-functions): here, returning a reference out of a function, from a conditional.
+
+```rust=
+fn get_default<'r, K: Hash + Eq + Copy, V: Default>(
+    map: &'r mut HashMap<K, V>,
+    key: K,
+) -> &'r mut V {
+    match map.get_mut(&key) { // -------------+ 'r
+        Some(value) => value,              // |
+        None => {                          // |
+            map.insert(key, V::default()); // |
+            //  ^~~~~~ ERROR               // |
+            map.get_mut(&key).unwrap()     // |
+        }                                  // |
+    }                                      // |
+}                                          // v
+```
+
+Returning the mutable reference `value` in the `Some` path requires the mutable loan on `map` to live until the end of the function. This prevents mutation in the `None` path even though no mutable loan on `map` would exist there in the first place. 
+
+Fixing this borrowck issue requires more precision about flow-sensitivity. It also hints at limitations in our modeling of lifetimes, which appear more clearly in cases with only slightly more complicated control flow, like [issue #47680](https://github.com/rust-lang/rust/issues/47680):
+
+```rust=
+struct Thing;
+
+impl Thing {
+    fn maybe_next(&mut self) -> Option<&mut Self> { None }
+}
+
+fn main() {
+    let mut temp = &mut Thing;
+
+    loop {
+        match temp.maybe_next() {
+            Some(v) => { temp = v; }
+            None => { }
+        }
+    }
+}
+```
+
+Here, the loan created on iteration N _may be used_ in the `None` path on iteration N+1, however the borrowed variable at that iteration is not the same.
+
+Issues like "NLL problem case #3", issue #47680 and others, were therefore deferred from NLLs, and left as future work, [Polonius](http://smallcultfollowing.com/babysteps/blog/2018/04/27/an-alias-based-formulation-of-the-borrow-checker/).
+
+The key ideas being:
+- switching from a model of _lifetimes_ as sets of points in the CFG (with _outlives_ relationships), to a model of _origins_ as sets of loans (with _subset_ relationships).
+- computing and tracking the subset relationships at each point in the Control Flow Graph (whereas the existing borrowck computes a single subtype relation).
+
+
+## Milestones
+
+This is a rough roadmap, where we have the most visibility on the first steps:
+- each step has unknowns that will define what things need to be done in the later steps
+- therefore we're talking more about milestones for the longer term roadmap, and proper tasks for the shorter term.
+
+Here are the roadmap's milestones:
+
+![Graph of the Polonius roadmap](../../../../images/inside-rust/2023-09-25-polonius-update/roadmap.png)
```

Comment:
This isn't rendering for me on GitHub. Unsure if this is anticipated (e.g. just a quirk of paths that work on the site but not inline) but wanted to note in case it was typo

Reply:
I don't believe images can render on github markdown previews because they are collected into a different web root `/static/` and this link pattern is used for resources in Inside Rust. 

Look at e.g. https://github.com/rust-lang/blog.rust-lang.org/pull/1103/files and https://github.com/rust-lang/blog.rust-lang.org/blob/92ac9e0679d4f41b338851aab1e18b33071bff9b/posts/inside-rust/2023-05-09-api-token-scopes.md?plain=1#L18 

FWIW the image does show up when locally serving the blog.